src/detect-ikev1-key-exchange-payload.c

```diff
@@ -0,0 +1,172 @@
+
+    if (DetectSignatureSetAppProto(s, ALPROTO_IKEV1) < 0)
+        return -1;
+
+    return 0;
+}
+
+static InspectionBuffer *GetClientKeyExchangeData(DetectEngineThreadCtx *det_ctx,
+        const DetectEngineTransforms *transforms, Flow *_f,
+        const uint8_t _flow_flags, void *txv, const int list_id)
+{
+    InspectionBuffer *buffer = InspectionBufferGet(det_ctx, list_id);
+    if (buffer->inspect == NULL) {
+        const uint8_t *b = NULL;
+        uint32_t b_len = 0;
+
+        if (rs_ikev1_state_get_key_exchange(txv, &b, &b_len) != 1)
+            return NULL;
+        if (b == NULL || b_len == 0)
+            return NULL;
+
+        InspectionBufferSetup(buffer, b, b_len);
+        InspectionBufferApplyTransforms(buffer, transforms);
+    }
+
+    return buffer;
+}
+
+static InspectionBuffer *GetServerKeyExchangeData(DetectEngineThreadCtx *det_ctx,
+                                          const DetectEngineTransforms *transforms, Flow *_f,
+                                          const uint8_t _flow_flags, void *txv, const int list_id)
+{
+    InspectionBuffer *buffer = InspectionBufferGet(det_ctx, list_id);
+    if (buffer->inspect == NULL) {
+        const uint8_t *b = NULL;
+        uint32_t b_len = 0;
+
+        if (rs_ikev1_state_get_key_exchange(txv, &b, &b_len) != 1)
+            return NULL;
+        if (b == NULL || b_len == 0)
+            return NULL;
+
+        InspectionBufferSetup(buffer, b, b_len);
+        InspectionBufferApplyTransforms(buffer, transforms);
+    }
+
+    return buffer;
+}
+
+void DetectIkev1KeyExchangeRegister(void)
+{
+    // register client_key_exchange
+    sigmatch_table[DETECT_AL_IKEV1_CLIENT_KEY_EXCHANGE].name = KEYWORD_NAME_CLIENT_KEY_EXCHANGE;
+    sigmatch_table[DETECT_AL_IKEV1_CLIENT_KEY_EXCHANGE].url = "/rules/" KEYWORD_DOC_CLIENT_KEY_EXCHANGE
+    sigmatch_table[DETECT_AL_IKEV1_CLIENT_KEY_EXCHANGE].desc = "sticky buffer to match on the IKEv1 client_key_exchange_payload";
+    sigmatch_table[DETECT_AL_IKEV1_CLIENT_KEY_EXCHANGE].Setup = DetectClientKeyExchangeSetup;
+    sigmatch_table[DETECT_AL_IKEV1_CLIENT_KEY_EXCHANGE].flags |= SIGMATCH_NOOPT|SIGMATCH_INFO_STICKY_BUFFER;
+
+    DetectAppLayerInspectEngineRegister2(BUFFER_NAME_CLIENT_KEY_EXCHANGE, ALPROTO_IKEV1,
+            SIG_FLAG_TOSERVER, 3,
+            DetectEngineInspectBufferGeneric, GetClientKeyExchangeData);
+
+    DetectAppLayerMpmRegister2(BUFFER_NAME_CLIENT_KEY_EXCHANGE, SIG_FLAG_TOSERVER, 1,
+            PrefilterGenericMpmRegister, GetClientKeyExchangeData, ALPROTO_IKEV1,
+            3);
+
+    DetectBufferTypeSetDescriptionByName(BUFFER_NAME_CLIENT_KEY_EXCHANGE, BUFFER_DESC_CLIENT_KEY_EXCHANGE);
+
+    g_buffer_client_key_exchange_id = DetectBufferTypeGetByName(BUFFER_NAME_CLIENT_KEY_EXCHANGE);
+    SCLogDebug("registering " BUFFER_NAME_CLIENT_KEY_EXCHANGE " rule option");
+
+    // register server_key_exchange
+    sigmatch_table[DETECT_AL_IKEV1_SERVER_KEY_EXCHANGE].name = KEYWORD_NAME_SERVER_KEY_EXCHANGE;
```

Comment:
Should not we use the rule direction instead of two keywords ?

Reply:
Can you be more precise on this? Not sure I understand...

Reply:
Instead of having 2 keywords `ikev1.client_key_exchange_payload` and `ikev1.server_key_exchange_payload`, could we have just one keyword `ikev1.key_exchange_payload` ? And then rely on additional classic keyword`flow:to_client` to distinguish if needed ?

Reply:
You are right, this should work. Have to admit that i didn't know it.
...perhaps i should read a bit more in the manual ;)